nemo_aligner/models/mm/stable_diffusion/megatron_sdxl_draftp_model.py

```diff
@@ -0,0 +1,500 @@
+                else:
+                    return True
+
+            frozen_submodules = []
+            frozen_submodule_names = []
+            for name, module in model.named_modules():
+                if (
+                    isinstance(module, nn.Module)
+                    and list(module.parameters())
+                    and all(not param.requires_grad for param in module.parameters())
+                    and (_ignore_first_stage_quant(name))
+                    and (_donot_ignore_embedders(name))
+                    and (_donot_ignore_diffusion_model(name))
+                ):
+                    frozen_submodule_names.append(name)
+                    frozen_submodules.append(module)
+            return frozen_submodule_names, frozen_submodules
+
+        if self.use_fsdp:
+            """Top-evel FSDP model sharding"""
+            # Shard the top-level model hierarchically. We shard the strategy-unwrapped model not
+            # to lose the structure of non-FSDP wrapped parameters (e.g, embedding)
+            # TODO: Currently the main parameter data type is kept in fp32 (when O2=False). This needs to be
+            # extended to support lower precision main parameters.
+            frozen_submodule_names, frozen_submodules = find_frozen_submodules(self.model)
+            for submodule in frozen_submodule_names:
+                logging.debug(f"Ignoring state {submodule} in FSDP.")
+            self.trainer.strategy.kwargs["ignored_states"] = frozen_submodules
+            # FSDP requires uniform status of require_grads
+            # Diffusion models like SD has frozen parts and needs to be added to 'ignored_states' from sharding for FSDP to work
+            self.model = self.trainer.strategy._setup_model(self.model)
+            # self.model.diffusion_model.first_stage = FullyShardedDataParallel
+            # Move the CPU-initialized model (with `use_cpu_initialization=True`) to GPU, which is to avoid
+            # out-of-memory carash before sharding. In case of GPU-initialized model, this is no-op.
+            self.model = self.model.cuda(torch.cuda.current_device())
+
+    def finish_inference(self):
+        return
+
+    def finish_training_step(self):
+        grad_reductions(self)
+
+    def infer(self, batch):
+        return
+
+    def prepare_for_inference(self):
+        return
+
+    def prepare_for_training_step(self):
+        # custom trainers will always zero grad for us
+        prepare_for_training_step(self.model, zero_grad=False)
+
+    def prepare_for_validation_step(self):
+        """things to call to prepare for validation
+        """
+        prepare_for_validation_step(self)
+        gbs = int(self.cfg.global_batch_size)
+        mbs = int(self.cfg.micro_batch_size)
+        dp_size = int(parallel_state.get_data_parallel_world_size())
+        configure_batch_sizes(mbs=mbs, gbs=gbs, dp=dp_size)
+
+    def finish_validation_step(self):
+        """things to call to prepare for validation
```

Comment:
This comment doesn't reflect method name

Reply:
I actually copied this from `megatron_sd_draftp_model`. I'll add the correct comment in both files.